runtime/jvmti/jvmtiHook.c

```diff
@@ -2231,26 +2231,40 @@ jvmtiHookGCCycleEnd(J9HookInterface** hook, UDATA eventNum, void* eventData, voi
 }
 
 /**
- * Determines if the class described by <tt>classLoadData</tt> can 
+ * Determines if the class described by <tt>data</tt> can 
  * be instrumented.
- * @param classLoadData 
+ * @param data The class file load hook event data
  * @return TRUE if the class can be instrumented. 
  */
 static BOOLEAN
-canClassBeInstrumented(J9VMClassLoadHookEvent* classLoadData)
-{
-	BOOLEAN result = TRUE;
-
-	/* Skip any class that does not have a name specified */
-	if (classLoadData->className == NULL) {
-		result = FALSE;
-	} else if (strcmp(classLoadData->className, "java/lang/J9VMInternals") == 0) {
-		result = FALSE;
-	} else if (strcmp(classLoadData->className, "java/lang/invoke/VarHandleInternal") == 0) {
-		result = FALSE;
+canClassBeInstrumented(J9VMClassLoadHookEvent* data)
+{
+	BOOLEAN modifiable = TRUE;
+	/* Only clasess on the bootpath may be marked unmodifiable */
+	if (data->classLoader == data->currentThread->javaVM->systemClassLoader) {
+		U_8 *classData = data->classData;
+		UDATA classDataLength =  data->classDataLength;
+		U_8 const annotationLength = LITERAL_STRLEN(J9_UNMODIFIABLE_CLASS_ANNOTATION);
+		if (classDataLength > (annotationLength + 4)) {
+			/* Make sure this at least looks like a class file */
+			if ((classData[0] = 0xCA) && (classData[1] = 0xFE) && (classData[2] = 0xBA) && (classData[3] = 0xBE)) {
+				U_8 *dataEnd = classData + classDataLength - annotationLength;
+				classData += 4;
+				while (classData < dataEnd) {
+					/* Look for u8 tag (1 for UTF8) followed by u16 string length */
+					if ((1 == classData[0]) && (0 == classData[1]) && (annotationLength == classData[2])) {
+						/* Now look for the annotation name */
+						if (0 == memcmp(classData + 3, J9_UNMODIFIABLE_CLASS_ANNOTATION, annotationLength)) {
+							modifiable = FALSE;
+							break;
+						}
```

Comment:
I think an extra `classData += 1;` can be done if we made into the if block as the block is looking for `{ 1 0 length }`

Reply:
Given that the string is short (and will not contain any matching substrings), I could completely optimize this by not using memcmp, and updating classData with each character compared.

Reply:
Sounds like a good experiment - perf test both versions and see which is actually faster.

Reply:
See below.